src/picongpu/include/plugins/ChargeConservation.tpp

```diff
@@ -0,0 +1,201 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "math/vector/Int.hpp"
+#include "math/vector/Float.hpp"
+#include "math/vector/Size_t.hpp"
+#include "math/vector/math_functor/abs.hpp"
+#include "math/vector/math_functor/max.hpp"
+#include "cuSTL/container/PseudoBuffer.hpp"
+#include "dataManagement/DataConnector.hpp"
+#include "fields/FieldJ.hpp"
+#include "math/Vector.hpp"
+#include "cuSTL/algorithm/mpi/Gather.hpp"
+#include "cuSTL/algorithm/mpi/Reduce.hpp"
+#include "cuSTL/container/DeviceBuffer.hpp"
+#include "cuSTL/container/HostBuffer.hpp"
+#include "cuSTL/algorithm/kernel/Foreach.hpp"
+#include "cuSTL/algorithm/host/Foreach.hpp"
+#include "cuSTL/cursor/NestedCursor.hpp"
+#include "lambda/Expression.hpp"
+#include <sstream>
+#include "algorithms/ForEach.hpp"
+#include "cuSTL/algorithm/kernel/Reduce.hpp"
+#include "nvidia/functors/Add.hpp"
+
+namespace picongpu
+{
+
+ChargeConservation::ChargeConservation()
+    : name("print the maximum charge deviation between particles and div E"),
+      prefix("ChargeConservation")
+{
+    Environment<>::get().PluginConnector().registerPlugin(this);
+}
+
+void ChargeConservation::pluginRegisterHelp(po::options_description& desc)
+{
+    desc.add_options()
+        ((this->prefix + "_frequency").c_str(),
+        po::value<uint32_t > (&this->notifyFrequency)->default_value(0), "notifyFrequency");
+}
+
+std::string ChargeConservation::pluginGetName() const {return this->name;}
+
+void ChargeConservation::pluginLoad()
+{
+    Environment<>::get().PluginConnector().setNotificationPeriod(this, this->notifyFrequency);
+}
+
+namespace detail
+{
+
+/**
+ * @class Div
+ * @brief divergence functor for 2D and 3D
+ *
+ * NOTE: This functor uses a Yee-cell stencil.
+ */
+template<int dim, typename ValueType>
+struct Div;
+
+template<typename ValueType>
+struct Div<DIM3, ValueType>
+{
+    typedef ValueType result_type;
+
+    template<typename Field>
+    HDINLINE ValueType operator()(Field field) const
+    {
+        const ValueType reciWidth = float_X(1.0) / CELL_WIDTH;
+        const ValueType reciHeight = float_X(1.0) / CELL_HEIGHT;
+        const ValueType reciDepth = float_X(1.0) / CELL_DEPTH;
+        return ((*field).x() - (*field(-1,0,0)).x()) * reciWidth +
+               ((*field).y() - (*field(0,-1,0)).y()) * reciHeight +
+               ((*field).z() - (*field(0,0,-1)).z()) * reciDepth;
+    }
+};
+
+template<typename ValueType>
+struct Div<DIM2, ValueType>
+{
+    typedef ValueType result_type;
+
+    template<typename Field>
+    HDINLINE ValueType operator()(Field field) const
+    {
+        const ValueType reciWidth = float_X(1.0) / CELL_WIDTH;
+        const ValueType reciHeight = float_X(1.0) / CELL_HEIGHT;
+        return ((*field).x() - (*field(-1,0)).x()) * reciWidth +
+               ((*field).y() - (*field(0,-1)).y()) * reciHeight;
```

Comment:
Ahh I see, this is the Yee curl. Right?
If this is the Yee curl, why we not use them? Than this should also be compatible with the Lehe field solver.